### PR TITLE
Clean up build.py

### PR DESCRIPTION
**Pull request by jchen351**
### Description
Clean up build.py

### Motivation and Context
Moving nonessential functions to the onnxruntime/tools/python/util package. Only left function that are meaningful to the build process.

### CODE REVIEW

File: tools/ci_build/build.py [hunk 1]
```diff
@@ -844,15 +733,6 @@ def update_submodules(source_dir):
     run_subprocess(["git", "submodule", "update", "--init", "--recursive"], cwd=source_dir)
 
 
-def is_docker():
```
Comment (jchen351) [inline tools/ci_build/build.py:847]: This function is no-longer needed

File: tools/ci_build/build.py [hunk 2]
```diff
@@ -14,118 +14,41 @@
...
-    for s in version.split("."):
-        with contextlib.suppress(ValueError):
-            v.append(int(s))
-    return tuple(v)
-
-
 SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
 REPO_DIR = os.path.normpath(os.path.join(SCRIPT_DIR, "..", ".."))
 
 sys.path.insert(0, os.path.join(REPO_DIR, "tools", "python"))
```
Comment (justinchuby) [inline tools/ci_build/build.py:20]: Modifying sys.path is usually not the best thing to do. Given a chance I would find ways to remove it.

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -14,118 +14,41 @@
...
 
-
+# The Following uil packages is from onnxruntime/tools/python/util
```
Comment (justinchuby) [inline tools/ci_build/build.py:22]:
```suggestion
# The Following utility packages is from onnxruntime/tools/python/util
```

File: tools/python/util/BuildError.py
Comment (justinchuby) [inline tools/python/util/BuildError.py:1]: Use snake case for module names

File: tools/python/util/BaseError.py
Comment (justinchuby) [inline tools/python/util/BaseError.py:1]: Snake case for file name

File: .gitignore [hunk 4]
```diff
@@ -196,3 +196,4 @@ Package.resolved
 .build/
 .swiftpm/
 repros/
+!/tools/ci_build/build_uitl/
```
Comment (justinchuby) [inline .gitignore:199]: Typo?

**Review by justinchuby — DISMISSED** [review]
LGTM with minor comments. I would also fix lint errors.

File: tools/python/util/build_helpers.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        return False
+    if args.ios and is_macOS():
+        return False
+    SYSTEM_COLLECTIONURI = os.getenv("SYSTEM_COLLECTIONURI")  # noqa: N806
+    if SYSTEM_COLLECTIONURI and SYSTEM_COLLECTIONURI != "https://dev.azure.com/onnxruntime/":
+        return False
+    return True
+
+
+def add_default_definition(definition_list, key, default_value):
```
Comment (justinchuby) [inline tools/python/util/build_helpers.py:71]: CodeQL: Mixing implicit and explicit returns may indicate an error as implicit returns always return None.

File: tools/python/util/build_helpers.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+def number_of_parallel_jobs(args):
+    return os.cpu_count() if args.parallel == 0 else args.parallel
+
+
+def number_of_nvcc_threads(args):
+    if args.nvcc_threads >= 0:
+        return args.nvcc_threads
+
+    nvcc_threads = 1
+    import psutil
```
Comment (justinchuby) [inline tools/python/util/build_helpers.py:91]: Prefer moving imports to the top of the module.

File: tools/python/util/build_helpers.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    try:
+        available_memory = psutil.virtual_memory().available
```
Comment (justinchuby) [inline tools/python/util/build_helpers.py:93]:
```suggestion
    
    try:
        import psutil
    except ImportError:
        print(
            "Failed to import psutil. Please `pip install psutil` for better estimation of nvcc threads. Use "
            "nvcc_threads=1"
        )
   
   available_memory = psutil.virtual_memory().available
```

I would keep the try-catch block small. Previously the import was not in the block so the catch was not meaningful.

File: tools/python/util/open_vino_utils.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            lib_path = os.path.join(os.path.dirname(__file__), lib)
+            if os.path.isdir(lib_path):
+                os.environ["PATH"] = os.path.abspath(lib_path) + ";" + os.environ["PATH"]
+                os.add_dll_directory(os.path.abspath(lib_path))
+
+
+def openvino_verify_device_type(device_read):
+    choices = ["CPU_FP32", "CPU_FP16", "GPU_FP32", "GPU_FP16"]
+
+    choices1 = [
```
Comment (justinchuby) [inline tools/python/util/open_vino_utils.py:41]: nit: I understand this is just moving code around - if possible, consider a more meaningful variable name?

File: tools/python/util/usage_error.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#  // Copyright (c) Microsoft Corporation. All rights reserved.
+#  // Licensed under the MIT License.
+from .base_error import BaseError
+
+
+class UsageError(BaseError):
```
Comment (justinchuby) [inline tools/python/util/usage_error.py:6]: Was there a motivation for keeping the errors in different modules? Looks like they can all be defined in the same module.

File: tools/python/util/usage_error.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+    """Usage related error."""
+
+    def __init__(self, message):
+        super().__init__(message)
```
Comment (justinchuby) [inline tools/python/util/usage_error.py:10]:
```suggestion
```

This should be redundant

**Comment by justinchuby** [comment]
An import error seen: https://github.com/microsoft/onnxruntime/actions/runs/7547312086/job/20546973767?pr=19156#step:5:12

**Comment by edgchen1** [comment]
> Moving nonessential functions to the onnxruntime/tools/python/util package. Only left function that are meaningful to the build process.

Can you clarify what "nonessential" means? I gather it's different from "unused", but it seems that anything that is used is "essential" in some way. I ask so that when we add new functions in the future we know where to put them.

**Comment by jchen351** [comment]
> > Moving nonessential functions to the onnxruntime/tools/python/util package. Only left function that are meaningful to the build process.
> 
> Can you clarify what "nonessential" means? I gather it's different from "unused", but it seems that anything that is used is "essential" in some way. I ask so that when we add new functions in the future we know where to put them.

We can put all future helper functions into the files from `tools/python/util/` folder. The build.py has been growing too large to navigate IMO.

**Review by justinchuby — DISMISSED** [review]
LGTM w/ lint errors fixed

**Comment by tianleiwu** [comment]
@jchen351, Is it still needed? If not, please close it.

**Pull request closed without merging**